### PR TITLE
 TEN-162 added in enum for enrolments for app code, added in throttle…

### PR DESCRIPTION
**Pull request by ChrisHillCap**
… action an service and some tests, added in app config value for percentage. not added actions to any controller

https://github.com/hmrc/app-config-qa/pull/15060
https://github.com/hmrc/app-config-staging/pull/10473
https://github.com/hmrc/app-config-production/pull/14607

### CODE REVIEW

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      case n if n < 0 => throw new IllegalArgumentException(s"percentage to throttle must be greater than 0, current percentage: $percentageToThrottle")
+      case _ if nino.length != 9 => throw new IllegalArgumentException(s"nino is incorrect length ${nino.length}")
+      case n => Try(nino.substring(6,8).toInt).map(ninoNumber => ninoNumber <= n).getOrElse(throw new IllegalArgumentException(s"nino was not valid format for throttle"))
+    }
+  }
+
+    private[services] def shouldAccountTypeBeThrottled(accountType: AccountTypes.Value,
+                                           percentageToThrottle: Int,
+                                           nino: String): Boolean = {
+    accountType != SINGLE_ACCOUNT && isNinoWithinThrottleThreshold(nino, percentageToThrottle)
```
Comment (Chrissystk) [inline app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala:67]: The account types that should be excluded from the throttle should also include PT_ASSIGNED_TO_CURRENT_USER and PT_ASSIGNED_TO_OTHER_USER

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+sealed trait ThrottleResult
+case object ThrottleApplied extends ThrottleResult
+case object ThrottleDoesNotApply extends ThrottleResult
+
+class ThrottlingService @Inject()(legacyAuthConnector: LegacyAuthConnector, appConfig: AppConfig) {
+
+  def throttle(accountType: AccountTypes.Value, nino: String, currentEnrolments: Set[Enrolment])(
```
Comment (Chrissystk) [inline app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala:39]: Should we not be saving the throttle result to cache? Then as a first step see if it is in the cache and if not then do the logic?

File: conf/application.conf [hunk 3]
```diff
@@ -124,7 +124,9 @@ microservice {
     }
   }
 }
-
+throttle {
+percentage = 100
+}
```
Comment (Chrissystk) [inline conf/application.conf:129]: I think the plan was to have set locally a fixed value that does will not cause the current journey tests to fail but allow a new journey test to be added that will not be selected to go through our journey.

**Review by Chrissystk — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala (hunk 1 above)
Comment (ChrisHillCap) [reply app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala:67]: Im thinking, maybe it should just also exclude PT_ASSIGNED_TO_CURRENT_USER
we want to include all other account types for now so it doesn't do exactly what we didn't want it to do, show the page for one NINO but not another.

File: conf/application.conf (hunk 3 above)
Comment (ChrisHillCap) [reply conf/application.conf:129]: arbirtrary value for now, but should be set to 0 in next task so only NINOS QQ112200Q are inlcuded

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala (hunk 2 above)
Comment (ChrisHillCap) [reply app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala:39]: the check of PT_ASSIGNED_TO_CURRENT_USER should catch this scenario, I'm unsure how much value a call to mongo gives us as its got a TTL and we wont be doing any other calls to AUTH IF user is PT_ASSIGNED_TO_CURRENT_USER

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala (hunk 1 above)
Comment (ChrisHillCap) [reply app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala:67]: ill make this change, good spot

Comment (ChrisHillCap) [reply app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala]: done

Comment (Chrissystk) [reply app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala]: No it needs to include both PT_ASSIGNED_CURRENT_USER and PT_ASSIGNED_TO_OTHER_USER. The page around PT enrolment already being assigned we always want to show. I have had discussions with Kit Pimm around this and he agreed this should be the case. The only way the PT_ASSIGNED_TO_OTHER_USER can be displayed is if a user has been through the silent enrolment (either through us or the background update process) or completed the multiple accounts no enrolments or multiple accounts with SA on one of the accounts. For people with multiple accounts they will automatically be within the throttle, however the single account holders could potentially create a new account and then be allowed in if this page is throttled which we don't want. Once the enrolment has been assigned for real the user needs to sign in with that account or report the account.

Comment (ChrisHillCap) [reply app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala]: ok I see this was that edge case Kit mentioned

Comment (Chrissystk) [reply app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala]: Yes it is also the same reason the background process can't go first :-)

**Review by Chrissystk — APPROVED** [review]

**Pull request merged**